go/database/mpt/verification_proof.go

```diff
@@ -0,0 +1,279 @@
+	}
+
+	if from < 0 {
+		from = 0
+		log.Printf("adjusting to a minimum block height: %d", from)
+	}
+
+	if to <= 0 || to > int(blockHeight) {
+		to = int(blockHeight) + 1
+		log.Printf("adjusting to a maximum block height: %d", to)
+	}
+
+	observer.Progress(fmt.Sprintf("Verifying total %d blocks", to-from))
+	for i := uint64(from); i < uint64(to); i++ {
+		observer.Progress(fmt.Sprintf("Collecting state for block... \t %d", i))
+		// collect all accounts for this block
+		visitor := accountCollectingVisitor{accounts: accountsMap{}, ctx: ctx}
+		if err := trie.VisitTrie(i, &visitor); err != nil {
+			observer.EndVerification(err)
+			return err
+		}
+		if interrupt.IsCancelled(ctx) {
+			return interrupt.ErrCanceled
+		}
+
+		observer.Progress(fmt.Sprintf("Collecting proof for block.. \t %d \t Accounts: %d", i, len(visitor.accounts)))
+		// collect all proofs for this block
+		rawProof := make(map[string]struct{})
+		for addr, data := range visitor.accounts {
+			if interrupt.IsCancelled(ctx) {
+				return interrupt.ErrCanceled
+			}
+
+			accountProof, err := trie.CreateWitnessProof(i, addr, maps.Keys(data.storage)...)
+			if err != nil {
+				observer.EndVerification(err)
+				return err
+			}
+			for _, element := range accountProof.GetElements() {
+				rawProof[element] = struct{}{}
+			}
+		}
+
+		rootHash, err := trie.GetHash(i)
+		if err != nil {
+			observer.EndVerification(err)
+			return err
+		}
+
+		observer.Progress(fmt.Sprintf("Recovering proof for block... \t %d", i))
+		// match proofs and database data
+		proof := CreateWitnessProofFromNodes(maps.Keys(rawProof))
+		observer.Progress(fmt.Sprintf("Verifying proof for block... \t %d", i))
+		if err := verifyProofState(ctx, rootHash, proof, visitor.accounts); err != nil {
+			observer.EndVerification(err)
+			return err
+		}
+	}
+
+	observer.EndVerification(nil)
+	return nil
+}
+
+// VerifyProofLiveTrie verifies the consistency of witness proofs for an live trie.
+// It reads the trie for the head block and loads all accounts and storage slots.
+// Then, it extracts witness proofs for the same accounts and storage slots.
+// Finally, the actual values from the trie are compared with the values extracted from the witness proofs.
+func VerifyProofLiveTrie(ctx context.Context, dir string, config MptConfig, observer VerificationObserver) error {
+	trie, err := OpenFileLiveTrie(dir, config, NodeCacheConfig{})
+	if err != nil {
+		return err
+	}
+
+	return errors.Join(
+		verifyProofLiveTrie(ctx, trie, observer),
+		trie.Close())
+}
+
+// verifyProofLiveTrie verifies the consistency of witness proofs for an live trie.
+// It reads the trie for the head block and loads all accounts and storage slots.
+// Then, it extracts witness proofs for the same accounts and storage slots.
+// Finally, the actual values from the trie are compared with the values extracted from the witness proofs.
+func verifyProofLiveTrie(ctx context.Context, trie *LiveTrie, observer VerificationObserver) error {
```

Comment:
The two methods `verifyProofLiveTrie` and `verifyProofArchiveTrie` have lots of code in common. The main difference is the extra loop over the blocks and the input type.

It would be great if the per-block verification code of `verifyProofArchiveTrie` could make use of the same code as the `verifyProofLiveTrie` function.

Reply:
right, unified so that it gets `*LiveTrie` from `*ArchiveTrie` and calls the method verifying the live trie